src/Raven.Server/Rachis/Leader.cs

```diff
@@ -740,12 +742,23 @@ public void Dispose()
                     ae.Execute(ambasaddor.Value.Dispose);
                 }
 
-
-                _newEntry.Dispose();
-                _voterResponded.Dispose();
-                _promotableUpdated.Dispose();
-                _shutdownRequested.Dispose();
-                _noop.Dispose();
+                var eventsLockTaken = false;
+                try
+                {
+                    //We are trying to modify the topology during the dispose, we will spin-wait.
+                    while (eventsLockTaken == false)
+                        Monitor.TryEnter(_eventsLocker, ref eventsLockTaken);
```

Comment:
Why do this busy wait here, and not just wait?

Reply:
because it is not expected to take too long, just wait for the events to fire.

Reply:
It isn't expected now, but it might be in the future, and we want to avoid such behavior from the get go